File: algorithm/프로그래머스/2/155651. 호텔 대실/호텔 대실.go
import (
    "sort"
    "strconv"
    "strings"
)

func solution(book_time [][]string) int {
    bookTime := make([][]int, len(book_time))
    for i, book := range book_time {
        start, end := book[0], book[1]
        
        sSplit := strings.Split(start, ":")
        sHour, _ := strconv.Atoi(sSplit[0])
        sEnd, _ := strconv.Atoi(sSplit[1])
        
        eSplit := strings.Split(end, ":")
        eHour, _ := strconv.Atoi(eSplit[0])
        eEnd, _ := strconv.Atoi(eSplit[1])
        
        bookTime[i] = []int{60 * sHour + sEnd, 60 * eHour + eEnd + 10}
    }
    
    sort.Slice(bookTime, func(i, j int) bool { return bookTime[i][0] < bookTime[j][0] })
    
    answer := 0
    var rooms [][]int
    for _, time := range bookTime {
        var check bool
        
        for i, room := range rooms {
            if room[1] <= time[0] {
                rooms[i] = time
                check = true
                break
            }
        }
        
        if !check {
            rooms = append(rooms, time)
        }
        
        if answer < len(rooms) {
            answer = len(rooms)
        }
    }
    
    return answer
}

File: algorithm/프로그래머스/2/155651. 호텔 대실/호텔 대실.py
def solution(book_time):
    answer = 0
    
    for i in range(len(book_time)):
        start, end = book_time[i]
        s_h, s_m = map(int, start.split(":"))
        s = 60 * s_h + s_m
        
        e_h, e_m = map(int, end.split(":"))
        e = 60 * e_h + e_m + 10
        
        book_time[i] = [s, e]
    
    sorted_book_time = sorted(book_time, key=lambda x: (x[0], x[1]))
    
    q = []
    for time in sorted_book_time:
        booked = False
        for i, room in enumerate(q):
            if room[1] <= time[0]:
                q[i] = time
                booked = True
                break
                
        if booked is False:
            q.append(time)
        
        answer = max(answer, len(q))
    
    return answer